reuse.go

```diff
@@ -0,0 +1,140 @@
+package libp2pquic
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/vishvananda/netlink"
+)
+
+type reuseConn struct {
+	net.PacketConn
+	refCount int32 // to be used as an atomic
+}
+
+func newReuseConn(conn net.PacketConn) *reuseConn {
+	return &reuseConn{PacketConn: conn}
+}
+
+func (c *reuseConn) IncreaseCount() { atomic.AddInt32(&c.refCount, 1) }
+func (c *reuseConn) DecreaseCount() { atomic.AddInt32(&c.refCount, -1) }
+func (c *reuseConn) GetCount() int  { return int(atomic.LoadInt32(&c.refCount)) }
+
+type reuse struct {
+	mutex sync.Mutex
+
+	unicast map[string] /* IP.String() */ map[int] /* port */ *reuseConn
```

Comment:
other nit: do we really want to map port to conn or just `map[*reuseConn]struct{}`?

Reply:
It needs to be a map of port to conn, if we ever want to reuse a connection that we dialed on for listening. I admit, this is a rather rare case, but it doesn't cost us anything.

Reply:
I'm not sure I understand what you mean by `string(IP)`.

Reply:
> https://github.com/libp2p/go-yamux/pull/6#pullrequestreview-270734844

:+1:

> I'm not sure I understand what you mean by string(IP).

That is, we can convert the raw bytes directly to a string rather than formatting as "xxx.xxx.xxx.xxx"). It would make lookups zero-allocation (`mymap[string(someByteArray)]` doesn't allocate).

Reply:
That's great! I wasn't aware that `net.IP` is just a `[]byte`.

Reply:
Great in theory, that is. It looks like there's no obvious way to normalize IP addresses, and it's failing in our tests. 

I managed to reproduce the issue locally, the IP is 192.168.46.226.
The kernel resolves this to: `net.IP{0xc0, 0xa8, 0x2e, 0xe2}`.
After listening, the local address is: `net.IP{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 0xc0, 0xa8, 0x2e, 0xe2}`

Normalizing this looks straightforward in this case, however, I'm a bit wary of pitfalls when using IPv6 addresses. Map lookups will hardly be the most expensive operation when starting a new QUIC listener / dialer, so I think we'll be fine for now with leaving this as a `net.IP.String()`.